File: Top-Interview-Questions/Tree/104-maximum-depth-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 * Runtime: 124 ms, faster than 5.14% of JavaScript online submissions for Maximum Depth of Binary Tree.
 * Memory Usage: 38.4 MB, less than 13.60% of JavaScript online submissions for Maximum Depth of Binary Tree.
 * TC: O(N) where N is the number of elements in root
 * SC: O(logN) for balanced and O(N) for extreme unbalanced where N is the number of elements in root
 *
 * If we solve this problem with BFS, then would the space complexity would be different?
 * Actually, the change in complexity is quite interesting! 
 * SC: O(1) for extreme unbalanced tree, O(N) for balanced tree
 */
var maxDepth = function(root) {
    if (!root) {
        return 0;
    }

    const leftDepth = maxDepth(root.left) + 1;
    const rightDepth = maxDepth(root.right) + 1;

    return Math.max(leftDepth, rightDepth);
};

/**
 * @param {TreeNode} root
 * @return {number}
 * Runtime: 100 ms, faster than 10.58% of JavaScript online submissions for Maximum Depth of Binary Tree.
 * Memory Usage: 41 MB, less than 5.10% of JavaScript online submissions for Maximum Depth of Binary Tree.
 */
var maxDepth = function(root) {
    if (!root) {
        return 0;
    }
    
    const queue = [];
    queue.unshift([root, 1]); // queue.add()
    
    let maxDepth = 0;
    while (queue.length !== 0) {
        const [node, depth] = queue.pop(); // queue.poll()
        maxDepth = Math.max(maxDepth, depth);
        
        if (node.left) {
            queue.unshift([node.left, depth + 1]);
        }
        
        if (node.right) {
            queue.unshift([node.right, depth + 1]);
        }
    }
    
    return maxDepth;
};

File: Top-Interview-Questions/Tree/104-maximum-depth-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// Solution #1: Using BFS
// Time Complexity: O(n) has to traverse all nodes
// Space Complexity: O(m) where m is max width of the tree
// Runtime: 12 ms, faster than 65.06% 
// Memory Usage: 19.6 MB, less than 5.16%
class Solution {
public:
    int maxDepth(TreeNode* root) {
            
        if(root == NULL)
            return 0;
        
        queue<TreeNode*> q;
        q.push(root);
        int nodeCount = q.size();
        int height = 0;
        
        while(1){
            
            nodeCount = q.size();
            
            if(nodeCount == 0)
                return height;
            
            height++;
            
            while(nodeCount > 0){
                TreeNode *node = new TreeNode();
                node = q.front();
                q.pop();
                
                if(node->left != NULL)
                    q.push(node->left);
                
                if(node->right != NULL)
                    q.push(node->right);
                
                nodeCount--;
            }
            
        }
        
    }
};

// Solution #1: Using DFS
// Time Complexity: O(n) has to traverse all nodes
// Space Complexity: O(h) where m is the max height of the tree
// Runtime: 12 ms, faster than 65.06%
// Memory Usage: 19.2 MB, less than 33.64%
class Solution {
public:
    int DFS(TreeNode *root){
        
        if(root == NULL)
            return 0;
            
        int leftHeight = DFS(root->left);
        int rightHeight = DFS(root->right);
        
        return max(leftHeight, rightHeight) + 1;
    }
    
    int maxDepth(TreeNode* root) {
        
        if(root == NULL)
            return 0;
        
        
        return DFS(root);  
    }
};